python/gtsam/examples/Pose2ISAM2Example.py

```diff
@@ -0,0 +1,155 @@
+
+See LICENSE for the license information
+
+Pose SLAM example using iSAM2 in the 2D plane.
+Author: Jerred Chen, Yusuf Ali
+Modelled after:
+    - VisualISAM2Example by: Duy-Nguyen Ta (C++), Frank Dellaert (Python)
+    - Pose2SLAMExample by: Alex Cunningham (C++), Kevin Deng & Frank Dellaert (Python)
+"""
+
+from __future__ import print_function
+import math
+import numpy as np
+from numpy.random import multivariate_normal as mult_gauss
+import gtsam
+import matplotlib.pyplot as plt
+import gtsam.utils.plot as gtsam_plot
+
+def Pose2SLAM_ISAM2_plot(graph, current_estimate):
+    """Plots incremental state of the robot for 2D Pose SLAM using iSAM2
+
+    Based on version by:
+        - Ellon Paiva (Python),
+        - Duy Nguyen Ta and Frank Dellaert (MATLAB)
+    """
+    marginals = gtsam.Marginals(graph, current_estimate)
+
+    fig = plt.figure(0)
+    axes = fig.gca()
+    plt.cla()
+
+    i = 1
+    while current_estimate.exists(i):
+        gtsam_plot.plot_pose2(0, current_estimate.atPose2(i), 0.5, marginals.marginalCovariance(i))
+        i += 1
+
+    plt.axis('equal')
+    axes.set_xlim(-1, 5)
+    axes.set_ylim(-1, 3)
+    plt.pause(1)
+    return marginals
+
+
+def Pose2SLAM_ISAM2_example():
+    """Perform 2D SLAM given the ground truth changes in pose as well as
+    simple loop closure detection.
+    """
+    plt.ion()
+
+    def vector3(x, y, z):
+        """Create 3d double numpy array."""
+        return np.array([x, y, z], dtype=float)
+
+    # Although this example only uses linear measurements and Gaussian noise models, it is important
+    # to note that iSAM2 can be utilized to its full potential during nonlinear optimization. This example
+    # simply showcases how iSAM2 may be applied to a Pose2 SLAM problem.
+    PRIOR_NOISE = gtsam.noiseModel.Diagonal.Sigmas(vector3(0.3, 0.3, 0.1))
+    ODOMETRY_NOISE = gtsam.noiseModel.Diagonal.Sigmas(vector3(0.2, 0.2, 0.1))
+
+    # Create a Nonlinear factor graph as well as the data structure to hold state estimates.
+    graph = gtsam.NonlinearFactorGraph()
+    initial_estimate = gtsam.Values()
+
+    # iSAM2 parameters which can adjust the threshold necessary to force relinearization and how many
+    # update calls are required to perform the relinearization.
+    parameters = gtsam.ISAM2Params()
+    parameters.setRelinearizeThreshold(0.1)
+    parameters.setRelinearizeSkip(1)
+    isam = gtsam.ISAM2(parameters)
+
+    # The ground truth odometry measurements (without noise) of the robot during the trajectory.
+    odom_arr = [(2, 0, 0),
+                (2, 0, math.pi/2),
+                (2, 0, math.pi/2),
+                (2, 0, math.pi/2),
+                (2, 0, math.pi/2)]
+
+    # The initial estimates for robot poses 2-5. Pose 1 is initialized by the prior.
+    # To demonstrate iSAM2 incremental optimization, poor initializations were intentionally inserted.
+    pose_est = [gtsam.Pose2(2.3, 0.1, -0.2),
```

Comment:
To make this example more realistic, you should initialize new variables incrementally using the measured odometry